middleware/authentication.go

```diff
@@ -23,3 +23,90 @@
 //
 // Source: https://pkg.go.dev/github.com/gorilla/mux?tab=doc#pkg-overview
 package middleware
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+
+	"github.com/lvl484/user-manager/logger"
+	"github.com/lvl484/user-manager/model"
+)
+
+const messageUnauthorized = "Authenticate failed"
+
+type UserProvider interface {
+	GetInfo(username string) (*model.User, error)
+}
+
+type BasicAuthentication struct {
+	ur UserProvider
+}
+
+func NewBasicAuthentication(ur UserProvider) *BasicAuthentication {
+	return &BasicAuthentication{ur: ur}
+}
+
+func (a *BasicAuthentication) Middleware(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, pass, ok := r.BasicAuth()
+		if !ok {
+			unauthorized(w)
+			return
+		}
+
+		userFromDB, err := a.ur.GetInfo(user)
+		if err != nil {
+			internalServerError(w, err)
+			return
+		}
+
+		matched, err := model.ComparePassword(pass, userFromDB.Password)
+		if err != nil {
+			internalServerError(w, err)
+			return
+		}
+
+		if !matched {
+			unauthorized(w)
+			return
+		}
+
+		logger.LogUM.Debugf("Authentication successful! Hello, %s", user)
+
+		handler.ServeHTTP(w, r)
+	})
+}
+
+func unauthorized(w http.ResponseWriter) {
```

Comment:
It would be great to move `Error` type and HTTP error handling to another package, for example: `http`, `web`, etc

Reply:
tried to move into package server, but received circular import error. What do you think if I move authentication middleware into server package?

Reply:
Let's move to `server/http/middleware/auth.go`

Reply:
Ok. I moved it